connect/transforms/src/main/java/org/apache/kafka/connect/transforms/Insert.java

```diff
@@ -0,0 +1,212 @@
+
+    private static class Config extends AbstractConfig {
+        public Config(Map<?, ?> originals) {
+            super(CONFIG_DEF, originals, false);
+        }
+    }
+
+    private static final class InsertionSpec {
+        final String name;
+        final boolean optional;
+
+        private InsertionSpec(String name, boolean optional) {
+            this.name = name;
+            this.optional = optional;
+        }
+
+        public static InsertionSpec parse(String spec) {
+            if (spec == null) return null;
+            if (spec.startsWith("?")) {
+                return new InsertionSpec(spec.substring(1), true);
+            }
+            if (spec.startsWith("!")) {
+                return new InsertionSpec(spec.substring(1), false);
+            }
+            return new InsertionSpec(spec, true);
+        }
+    }
+
+    private static final class SchemaUpdateCacheEntry {
+        final Schema base;
+        final Schema updated;
+
+        private SchemaUpdateCacheEntry(Schema base, Schema updated) {
+            this.base = base;
+            this.updated = updated;
+        }
+    }
+
+    private InsertionSpec topicField;
+    private InsertionSpec partitionField;
+    private InsertionSpec offsetField;
+    private InsertionSpec timestampField;
+    private boolean applicable;
+
+    // Expect this functionality to be refactored out as more transformations that operate on Schema's are added.
+    private final ThreadLocal<SchemaUpdateCacheEntry> schemaUpdateCache = new ThreadLocal<>();
+
+    @Override
+    public void init(Map<String, Object> props) {
+        final Config config = new Config(props);
+        topicField = InsertionSpec.parse(config.getString("topic"));
+        partitionField = InsertionSpec.parse(config.getString("partition"));
+        offsetField = InsertionSpec.parse(config.getString("offset"));
+        timestampField = InsertionSpec.parse(config.getString("timestamp"));
+        applicable = topicField != null || partitionField != null || offsetField != null || timestampField != null;
+    }
+
+    @Override
+    public R apply(R record) {
+        if (!applicable) return record;
+
+        final Schema schema = operatingSchema(record);
+        final Object value = operatingValue(record);
+
+        if (schema == null)
+            throw new DataException("Schemaless mode not supported by " + getClass().getName());
+
+        if (schema.type() != Schema.Type.STRUCT)
+            throw new DataException("Can only operate on Struct types: " + getClass().getName());
+
+        final Schema updatedSchema;
+        final SchemaUpdateCacheEntry schemaUpdateCacheEntry = schemaUpdateCache.get();
+        if (schemaUpdateCacheEntry == null || schemaUpdateCacheEntry.base != schema) {
+            updatedSchema = makeUpdatedSchema(schema);
+            schemaUpdateCache.set(new SchemaUpdateCacheEntry(schema, updatedSchema));
```

Comment:
Is the cache size of 1 what we want? Seems like it won't be great in cases where you're migrating schemas so you see them constantly changing. We have some cache utilities in o.a.k.common.cache that we could use for a larger cache.